src/rebar_compiler_erl.erl

```diff
@@ -245,20 +245,21 @@ module_to_erl(Mod) ->
     atom_to_list(Mod) ++ ".erl".
 
 parse_attrs(Fd, Includes, Dir) ->
-    case io:parse_erl_form(Fd, "") of
-        {ok, Form, _Line} ->
-            case erl_syntax:type(Form) of
-                attribute ->
-                    NewIncludes = process_attr(Form, Includes, Dir),
-                    parse_attrs(Fd, NewIncludes, Dir);
-                _ ->
-                    parse_attrs(Fd, Includes, Dir)
-            end;
-        {eof, _} ->
-            Includes;
-        _Err ->
-            parse_attrs(Fd, Includes, Dir)
-    end.
+    DupIncludes = case io:parse_erl_form(Fd, "") of
+                      {ok, Form, _Line} ->
+                          case erl_syntax:type(Form) of
+                              attribute ->
+                                  NewIncludes = process_attr(Form, Includes, Dir),
+                                  parse_attrs(Fd, NewIncludes, Dir);
+                              _ ->
+                                  parse_attrs(Fd, Includes, Dir)
+                          end;
+                      {eof, _} ->
+                          Includes;
```

Comment:
Since this is a recursive function, you'll end up re-sorting on every iteration ever by doing a `lists:usort(DupIncludes)` at the end of the function -- all clauses will hit it.

You'll notice though that this function is tail recursive, so there is only one terminal case that returns the last value, and that's the clause here on lines 257-258, with `{eof, _} -> Includes;`. This is where you should sort instead, so we end up with:

```
...
    {eof, _} ->
        lists:usort(Includes);
...
```
This will ensure that only the final iteration will sort, and the rest will leave as is.

Reply:
Good point